Core/ConnectedShapeBasedInterpolation.h

```diff
@@ -0,0 +1,220 @@
+
+#include <itkCastImageFilter.h>
+#include <itkTranslationTransform.h>
+#include <itkNearestNeighborInterpolateImageFunction.h>
+#include <itkImageRegionConstIteratorWithIndex.h>
+#include <itkResampleImageFilter.h>
+#include <itkSignedMaurerDistanceMapImageFilter.h>
+#include <itkExtractImageFilter.h>
+#include <itkBinaryThresholdImageFilter.h>
+#include <itkConnectedComponentImageFilter.h>
+
+#include "itkSliceContiguousImage.h"
+
+#include <vector>
+
+namespace iseg {
+
+class ConnectedShapeBasedInterpolation
+{
+public:
+	using image_type = itk::Image<float, 2>;
+	using mask_type = itk::Image<unsigned char, 2>;
+	using image_stack_type = itk::SliceContiguousImage<float>;
+
+	std::vector<image_type::Pointer> interpolate(const image_type* slice1, const image_type* slice2, int number_of_slices)
+	{
+		int num_objects1 = 0;
+		int num_objects2 = 0;
+		auto objects1 = get_components(slice1, num_objects1);
+		auto objects2 = get_components(slice2, num_objects2);
+		if (num_objects1 != 1 || num_objects2 != 1)
+		{
+			throw std::runtime_error("Detected multiple foreground objects");
+		}
+
+		auto x1 = get_center_of_mass(objects1, num_objects1)[0];
+		auto x2 = get_center_of_mass(objects2, num_objects2)[0];
+
+		itk::Vector<double, 2> translation;
+		translation[0] = x1[0] - x2[0];
+		translation[1] = x1[1] - x2[1];
+		auto slice2_moved = move_image(slice2, -translation);
+
+		auto sdf1 = compute_sdf(slice1);
+		auto sdf2 = compute_sdf(slice2_moved);
+
+		// slack to 3d image
+		using image_stack_type = itk::SliceContiguousImage<float>;
+		image_stack_type::IndexType start;
+		image_stack_type::SizeType size;
+		image_stack_type::SpacingType spacing;
+		image_stack_type::PointType origin;
+		for (int i = 0; i < 2; i++)
+		{
+			start[i] = slice1->GetBufferedRegion().GetIndex()[i];
+			size[i] = slice1->GetBufferedRegion().GetSize()[i];
+			spacing[i] = slice1->GetSpacing()[i];
+			origin[i] = slice1->GetOrigin()[i];
+		}
+		start[2] = 0;
+		size[2] = 2;
+		spacing[2] = 1.0; // TODO
+		origin[2] = 0.0;
+
+		auto image3d = image_stack_type::New();
+		image3d->SetRegions(image_stack_type::RegionType(start, size));
+		image3d->SetSpacing(spacing);
+		image3d->SetOrigin(origin);
+
+		std::vector<float*> slices;
+		slices.push_back(sdf1->GetPixelContainer()->GetImportPointer());
+		slices.push_back(sdf2->GetPixelContainer()->GetImportPointer());
+
+		auto container = image_stack_type::PixelContainer::New();
+		container->SetImportPointersForSlices(slices, size[0] * size[1], false);
+		image3d->SetPixelContainer(container);
+
+		std::vector<image_type::Pointer> interpolated_slices;
+		for (int i = 1; i <= number_of_slices; i++)
+		{
+			auto slice = interpolate(image3d, i / (number_of_slices + 1.0), translation);
+			interpolated_slices.push_back(slice);
+		}
+		return interpolated_slices;
+	}
+
+private:
+	/// find connected foreground regions (called "objects")
+	mask_type::Pointer get_components(const image_type* img, int& num_objects) const
```

Comment:
if itk::image are already smart pointers (which I think they are)... pass as const reference

Reply:
you mean return? I think I need to return a ::Pointer to keep it alive.

Reply:
no I meant the arguments. the return value is clear